add clang-proto-fuzzer to oss-fuzz

I am trying to add clang-proto-fuzzer to oss-fuzz (as part of the `llvm` project). 
This is the diff: 
```
--- a/projects/llvm/build.sh
+++ b/projects/llvm/build.sh
@@ -32,8 +32,10 @@ cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm \
     -DCMAKE_C_COMPILER=`pwd`/../build0/bin/clang \
     -DCMAKE_CXX_COMPILER=`pwd`/../build0/bin/clang++ \
     -DLLVM_USE_SANITIZE_COVERAGE=YES \
+    -DCLANG_ENABLE_PROTO_FUZZER=ON \
     -DLLVM_USE_SANITIZER=Address
 
-ninja clang-fuzzer
-
-cp bin/clang-fuzzer $OUT
+for fuzzer in clang-fuzzer clang-proto-fuzzer; do
+  ninja $fuzzer
+  cp bin/$fuzzer $OUT
+done

```

However, I get a build failure while configuring libprotobuf-mutator: 
```
[48/71] Linking CXX executable examples/libfuzzer/libfuzzer_example
FAILED: : && /src/build/../build0/bin/clang++   -fno-exceptions -Werror -Wall -O3 -DNDEBUG  -fsanitize-coverage=trace-pc-guard,trace-cmp examples/libfuzzer/CMakeFiles/libfuzzer_bin_example.dir/li
bfuzzer_bin_example.cc.o  -o examples/libfuzzer/libfuzzer_bin_example  examples/libfuzzer/libfuzzer-example-proto.a src/libfuzzer/libprotobuf-mutator-libfuzzer.a external.libfuzzer/src/external.l
ibfuzzer-build/libFuzzer.a src/libprotobuf-mutator.a /usr/local/lib/libprotobuf.so -pthread -Wl,-rpath,/usr/local/lib && :
examples/libfuzzer/libfuzzer-example-proto.a(libfuzzer_example.pb.cc.o): In function `libfuzzer_example::protobuf_libfuzzer_5fexample_2eproto::TableStruct::InitDefaultsImpl()':
libfuzzer_example.pb.cc:(.text+0x76): undefined reference to `google::protobuf::internal::fixed_address_empty_string[abi:cxx11]'
examples/libfuzzer/libfuzzer-example-proto.a(libfuzzer_example.pb.cc.o): In function `libfuzzer_example::protobuf_libfuzzer_5fexample_2eproto::AddDescriptorsImpl()':
libfuzzer_example.pb.cc:(.text+0x14d): undefined reference to `google::protobuf::MessageFactory::InternalRegisterGeneratedFile(char const*, void (*)(std::__cxx11::basic_string<char, std::char_tra
its<char>, std::allocator<char> > const&))'
examples/libfuzzer/libfuzzer-example-proto.a(libfuzzer_example.pb.cc.o): In function `libfuzzer_example::Msg::Msg()':
libfuzzer_example.pb.cc:(.text+0x290): undefined reference to `google::protobuf::internal::fixed_address_empty_string[abi:cxx11]'

```


Do we need to build `libfuzzer_example ` when building libprotobuf-mutator here? 

If yes, can we honor the CFLAGS provided by OSS-Fuzz? 
Looks like this build crashes because OSS-Fuzz wants libc++ instead of libstdc++.

--- a/projects/llvm/Dockerfile
+++ b/projects/llvm/Dockerfile
@@ -28,10 +28,4 @@ RUN svn co http://llvm.org/svn/llvm-project/llvm/trunk llvm > svn.log 2>&1
 RUN cd llvm/tools && svn co http://llvm.org/svn/llvm-project/cfe/trunk clang -r $(cd ../ && svn info | grep Revision | awk '{print $2}') >> svn.log 2>&1 
 RUN cd llvm/projects && svn co http://llvm.org/svn/llvm-project/compiler-rt/trunk compiler-rt -r $(cd ../ && svn info | grep Revision | awk '{print $2}')  >> svn.log 2>&1 
 
-# Build plain LLVM (stage 0)
-RUN mkdir build0
-RUN unset CC CXX CFLAGS CXXFLAGS; cd build0 && cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm
-RUN cd build0 && ninja clang compiler-rt
-RUN cd build0 && ninja fuzzer
-
 COPY build.sh $SRC/

--- a/projects/llvm/Dockerfile
+++ b/projects/llvm/Dockerfile
@@ -23,7 +23,6 @@ RUN apt-get install -y autoconf automake libtool curl make g++ unzip wget git \
 
 # Get protobuf
 RUN wget -qO- https://github.com/google/protobuf/releases/download/v3.3.0/protobuf-cpp-3.3.0.tar.gz | tar zxf -
-RUN cd protobuf-3.3.0 && ./autogen.sh && ./configure && make -j $(nproc) && make check -j $(nproc) && make install && ldconfig
 # Get LLVM
 RUN svn co http://llvm.org/svn/llvm-project/llvm/trunk llvm
 RUN cd llvm/tools && svn co http://llvm.org/svn/llvm-project/cfe/trunk clang -r $(cd ../ && svn info | grep Revision | awk '{print $2}')

--- a/projects/llvm/build.sh
+++ b/projects/llvm/build.sh
@@ -23,15 +23,25 @@
 # TODO(kcc): honor CFLAGS/CXXFLAGS to allow building with msan/ubsan
 
 # Build and install protobuf
-pushd protobuf-3.3.0
-./autogen.sh
-./configure --disable-shared
-make -j $(nproc)
-make check -j $(nproc)
-make install
-ldconfig
-popd
 
+build_protobuf() {
+  ./autogen.sh
+  ./configure --disable-shared
+  make -j $(nproc)
+  make check -j $(nproc)
+  make install
+  ldconfig
+}
+
+(cd protobuf-3.3.0 && build_protobuf)
+
+readonly FUZZERS=( \
+  clang-fuzzer \
+  clang-proto-fuzzer \
+  clang-format-fuzzer \
+  llvm-dwarfdump-fuzzer \
+  llvm-isel-fuzzer \
+)
 case $SANITIZER in
   address) LLVM_SANITIZER="Address" ;;
   undefined) LLVM_SANITIZER="Undefined" ;;
@@ -39,40 +49,21 @@ case $SANITIZER in
   *) LLVM_SANITIZER="" ;;
 esac
 
-mkdir cpf-build
-pushd cpf-build
+mkdir build
+cd build
 cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm \
     -DLLVM_ENABLE_ASSERTIONS=ON \
     -DCMAKE_C_COMPILER="${CC}" \
     -DCMAKE_CXX_COMPILER="${CXX}" \
     -DCMAKE_C_FLAGS="${CFLAGS}" \
     -DCMAKE_CXX_FLAGS="${CXXFLAGS}" \
-    -DLIB_FUZZING_ENGINE="${LIB_FUZZING_ENGINE}" \
+    -DLLVM_LIB_FUZZING_ENGINE="${LIB_FUZZING_ENGINE}" \
     -DCLANG_ENABLE_PROTO_FUZZER=ON \
     -DLLVM_USE_SANITIZER="${LLVM_SANITIZER}"
-for fuzzer in clang-fuzzer clang-proto-fuzzer; do
-  ninja $fuzzer
-  cp bin/$fuzzer $OUT
-done
-popd
-
-mkdir build
-cd build
-
-unset CC CXX CFLAGS CXXFLAGS
-
-cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm \
-    -DLLVM_ENABLE_ASSERTIONS=ON \
-    -DCMAKE_C_COMPILER=`pwd`/../build0/bin/clang \
-    -DCMAKE_CXX_COMPILER=`pwd`/../build0/bin/clang++ \
-    -DLLVM_NO_DEAD_STRIP=ON \
-    -DLLVM_USE_SANITIZE_COVERAGE=YES \
-    -DLLVM_USE_SANITIZER=Address
-
-for fuzzer in clang-format-fuzzer llvm-dwarfdump-fuzzer llvm-isel-fuzzer; do
+for fuzzer in "${FUZZERS[@]}"; do
   ninja $fuzzer
   cp bin/$fuzzer $OUT
 done
 
-# isel-fuzzer encodes its default fags in the name.
+# isel-fuzzer encodes its default flags in the name.
 mv $OUT/llvm-isel-fuzzer $OUT/llvm-isel-fuzzer=aarch64-gisel

--- a/projects/llvm/build.sh
+++ b/projects/llvm/build.sh
@@ -22,6 +22,40 @@
 # and LLVM's own copy of libFuzzer.
 # TODO(kcc): honor CFLAGS/CXXFLAGS to allow building with msan/ubsan
 
+# Build and install protobuf
+pushd protobuf-3.3.0
+./autogen.sh
+./configure --disable-shared
+make -j $(nproc)
+make check -j $(nproc)
+make install
+ldconfig
+popd
+
+case $SANITIZER in
+  address) LLVM_SANITIZER="Address" ;;
+  undefined) LLVM_SANITIZER="Undefined" ;;
+  memory) LLVM_SANITIZER="MemoryWithOrigins" ;;
+  *) LLVM_SANITIZER="" ;;
+esac
+
+mkdir cpf-build
+pushd cpf-build
+cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm \
+    -DLLVM_ENABLE_ASSERTIONS=ON \
+    -DCMAKE_C_COMPILER="${CC}" \
+    -DCMAKE_CXX_COMPILER="${CXX}" \
+    -DCMAKE_C_FLAGS="${CFLAGS}" \
+    -DCMAKE_CXX_FLAGS="${CXXFLAGS}" \
+    -DLIB_FUZZING_ENGINE="${LIB_FUZZING_ENGINE}" \
+    -DCLANG_ENABLE_PROTO_FUZZER=ON \
+    -DLLVM_USE_SANITIZER="${LLVM_SANITIZER}"
+for fuzzer in clang-fuzzer clang-proto-fuzzer; do
+  ninja $fuzzer
+  cp bin/$fuzzer $OUT
+done
+popd
+
 mkdir build
 cd build
 
@@ -34,7 +68,7 @@ cmake -GNinja -DCMAKE_BUILD_TYPE=Release ../llvm \
     -DLLVM_USE_SANITIZE_COVERAGE=YES \
     -DLLVM_USE_SANITIZER=Address
 
-for fuzzer in clang-fuzzer clang-format-fuzzer llvm-dwarfdump-fuzzer; do
+for fuzzer in clang-format-fuzzer llvm-dwarfdump-fuzzer; do
   ninja $fuzzer
   cp bin/$fuzzer $OUT
 done

